File: frontend/scripts/interview.js
import { createSession, getNextQuestion, evaluateAnswer, generateReport } from './api.js';
import { addAIMessage, addUserMessage, addTypingIndicator, removeTypingIndicator } from './components.js';

const sendButton = document.getElementById('send-button');
const userInput = document.getElementById('user-input');

let interviewState = {
    sessionId: null,
    currentQuestionId: null,
    isProcessing: false,
    isInterviewFinished: false,
};

// Disable input and button when processing
const toggleProcessing = (isProcessing) => {
    interviewState.isProcessing = isProcessing;
    sendButton.disabled = isProcessing;
    userInput.disabled = isProcessing;
    if (isProcessing) {
        addTypingIndicator();
    } else {
        removeTypingIndicator();
    }
};

export async function startInterview() {
    try {
        addTypingIndicator();
        const session = await createSession();
        interviewState.sessionId = session.session_id;

        removeTypingIndicator();
        addAIMessage("Hello, I am the AI Excel Interviewer. I will be conducting your technical assessment today. We will go through a series of questions. Your answers will be evaluated on correctness and efficiency. Let's begin. ");
        
        await new Promise(resolve => setTimeout(resolve, 2000));
        await askNextQuestion();

    } catch (error) {
        console.error("Failed to start interview:", error);
        removeTypingIndicator();
        addAIMessage("I'm sorry, I couldn't connect to the interview server. Please check the backend service. Error details are logged to the console.");
    }
}

export async function handleUserAnswer(userAnswer) {
    if (interviewState.isProcessing || interviewState.isInterviewFinished) {
        return;
    }

    addUserMessage(userAnswer);
    toggleProcessing(true);

    try {
        await evaluateAnswer(interviewState.sessionId, interviewState.currentQuestionId, userAnswer);
        await askNextQuestion();
    } catch (error) {
        console.error("Error during answer evaluation or next question fetch:", error);
        addAIMessage(`I'm sorry, an error occurred while processing your answer. Please check the browser console and the backend server logs for more details. Error: ${error.message}`);
        toggleProcessing(false);
    }
}

async function askNextQuestion() {
    toggleProcessing(true);
    try {
        const nextQuestionData = await getNextQuestion(interviewState.sessionId);
        toggleProcessing(false);

        if (nextQuestionData.is_last_question && nextQuestionData.question_id === 'end') {
            addAIMessage(nextQuestionData.question);
            interviewState.isInterviewFinished = true;
            await new Promise(resolve => setTimeout(resolve, 2000));
            await generateFinalReport();
        } else {
            addAIMessage(nextQuestionData.question);
            interviewState.currentQuestionId = nextQuestionData.question_id;
        }
    } catch (error) {
        console.error("Error fetching next question:", error);
        toggleProcessing(false);
        addAIMessage("I'm sorry, an error occurred while getting the next question. Please check the backend server and try again.");
    }
}

async function generateFinalReport() {
    toggleProcessing(true);
    try {
        const reportData = await generateReport(interviewState.sessionId);
        toggleProcessing(false);
        addAIMessage("Thank you for your time. Here is your final performance report:");
        addAIMessage(reportData.report);
    } catch (error) {
        console.error("Error generating report:", error);
        toggleProcessing(false);
        addAIMessage("I'm sorry, an error occurred while generating your report. Please try again later.");
    }
}

File: backend/api/routes/interview.py
from fastapi import APIRouter, HTTPException
from ..models.interview_models import QuestionResponse
from ..agents.interview_manager import InterviewManager
from ..utils.excel_knowledge import EXCEL_QUESTIONS
from ..services import state_manager

router = APIRouter()

@router.post("/start", response_model=QuestionResponse)
async def start_interview():
    """Starts a new interview session and returns the first question."""
    session_id = state_manager.create_session()
    first_question = EXCEL_QUESTIONS[0]
    state_manager.update_session_state(session_id, "current_question_id", first_question["id"])
    return {
        "question": "Hello, I am the AI Excel Interviewer. I will be conducting your technical assessment today. We will go through a series of questions. Your answers will be evaluated on correctness and efficiency. Let's begin.",
        "question_id": first_question["id"],
        "is_last_question": len(EXCEL_QUESTIONS) == 1,
        "session_id": session_id
    }

@router.get("/next-question/{session_id}", response_model=QuestionResponse)
async def get_next_question(session_id: str):
    """Returns the next question in the interview sequence."""
    session_state = state_manager.get_session_state(session_id)
    if not session_state:
        raise HTTPException(status_code=404, detail="Session not found.")
        
    current_question_id = session_state.get("current_question_id")
    interview_manager = InterviewManager()
    
    # Find the next question based on the current one
    current_index = next((i for i, q in enumerate(EXCEL_QUESTIONS) if q["id"] == current_question_id), -1)
    
    if current_index + 1 >= len(EXCEL_QUESTIONS):
        return {
            "question": "End of interview. Please proceed to generate the report.",
            "question_id": "end",
            "is_last_question": True,
        }

    next_question = EXCEL_QUESTIONS[current_index + 1]
    state_manager.update_session_state(session_id, "current_question_id", next_question["id"])

    return {
        "question": next_question["text"],
        "question_id": next_question["id"],
        "is_last_question": (current_index + 2) == len(EXCEL_QUESTIONS),
    }
